File: backend/models/kfa.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// KFAProductType represents the type of KFA product
type KFAProductType string

const (
	KFAProductTypeObat          KFAProductType = "obat"           // Obat
	KFAProductTypeAlatKesehatan KFAProductType = "alat_kesehatan" // Alat Kesehatan
)

// MedicineKFAMapping represents the mapping between local medicine and SatuSehat KFA code
type MedicineKFAMapping struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// Local Medicine Reference
	MedicineID uint      `gorm:"not null;uniqueIndex" json:"medicine_id"` // One medicine can only have one KFA mapping
	Medicine   *Medicine `gorm:"foreignKey:MedicineID" json:"medicine,omitempty"`

	// KFA (Kode Farmasi Indonesia) Information
	KFACode        string         `gorm:"not null;size:50;index" json:"kfa_code"`          // Kode KFA (e.g., "93001295")
	KFAName        string         `gorm:"not null;size:500" json:"kfa_name"`               // Nama di KFA
	KFAProductType KFAProductType `gorm:"not null;size:50;default:'obat'" json:"kfa_type"` // Tipe produk KFA
	KFAFarmalkes   string         `gorm:"size:100" json:"kfa_farmalkes"`                   // Kode Farmalkes (jika ada)

	// Additional KFA Info
	KFADisplayName  string `gorm:"size:500" json:"kfa_display_name"` // Display name untuk FHIR
	KFAForm         string `gorm:"size:100" json:"kfa_form"`         // Bentuk sediaan di KFA
	KFAStrength     string `gorm:"size:100" json:"kfa_strength"`     // Kekuatan di KFA
	KFAManufacturer string `gorm:"size:200" json:"kfa_manufacturer"` // Produsen di KFA

	// Verification
	IsVerified   bool       `gorm:"default:false" json:"is_verified"` // Sudah diverifikasi mapping-nya
	VerifiedAt   *time.Time `json:"verified_at,omitempty"`            // Waktu verifikasi
	VerifiedByID *uint      `json:"verified_by_id,omitempty"`         // User yang memverifikasi
	VerifiedBy   *User      `gorm:"foreignKey:VerifiedByID" json:"verified_by,omitempty"`

	// Status
	IsActive bool   `gorm:"default:true" json:"is_active"`
	Notes    string `gorm:"type:text" json:"notes"` // Catatan tambahan

	// SatuSehat Medication Resource
	SatuSehatMedicationID string     `gorm:"size:100;index" json:"satusehat_medication_id"` // ID Medication resource di SatuSehat
	SatuSehatSentAt       *time.Time `json:"satusehat_sent_at,omitempty"`                   // Waktu kirim ke SatuSehat
}

// TableName sets the table name for MedicineKFAMapping
func (MedicineKFAMapping) TableName() string {
	return "medicine_kfa_mappings"
}

// KFAMaster represents the master data of KFA codes (can be imported from SatuSehat)
type KFAMaster struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// KFA Information
	KFACode        string         `gorm:"not null;uniqueIndex;size:50" json:"kfa_code"` // Kode KFA unik
	KFAName        string         `gorm:"not null;size:500;index" json:"kfa_name"`      // Nama produk
	KFAProductType KFAProductType `gorm:"not null;size:50;default:'obat'" json:"kfa_type"`
	KFAFarmalkes   string         `gorm:"size:100" json:"kfa_farmalkes"` // Kode Farmalkes

	// Product Details
	Form         string `gorm:"size:100" json:"form"`         // Bentuk sediaan
	Strength     string `gorm:"size:100" json:"strength"`     // Kekuatan/Dosis
	Manufacturer string `gorm:"size:200" json:"manufacturer"` // Produsen
	Category     string `gorm:"size:100" json:"category"`     // Kategori obat

	// Status
	IsActive bool `gorm:"default:true" json:"is_active"`

	// SatuSehat Sync Info
	SyncedAt *time.Time `json:"synced_at,omitempty"` // Last sync from SatuSehat
}

// TableName sets the table name for KFAMaster
func (KFAMaster) TableName() string {
	return "kfa_masters"
}

File: frontend/src/lib/api/kfa.ts
import { api } from './client';

// ============================================================================
// KFA (Kode Farmasi Indonesia) Types
// ============================================================================

export type KFAProductType = 'obat' | 'alat_kesehatan';

export interface KFAMaster {
  id: number;
  kfa_code: string;
  kfa_name: string;
  kfa_type: KFAProductType;
  kfa_farmalkes: string;
  form: string;
  strength: string;
  manufacturer: string;
  category: string;
  is_active: boolean;
  synced_at?: string;
  created_at: string;
  updated_at: string;
}

export interface MedicineKFAMapping {
  id: number;
  medicine_id: number;
  medicine?: {
    id: number;
    code: string;
    name: string;
    generic_name: string;
    form: string;
    strength: string;
    unit: string;
    manufacturer: string;
    category: string;
    type: string;
  };
  kfa_code: string;
  kfa_name: string;
  kfa_type: KFAProductType;
  kfa_farmalkes: string;
  kfa_display_name: string;
  kfa_form: string;
  kfa_strength: string;
  kfa_manufacturer: string;
  is_verified: boolean;
  verified_at?: string;
  verified_by_id?: number;
  verified_by?: {
    id: number;
    full_name: string;
    username: string;
  };
  is_active: boolean;
  notes: string;
  satusehat_medication_id: string;
  satusehat_sent_at?: string;
  created_at: string;
  updated_at: string;
}

export interface KFAMappingStats {
  total_medicines: number;
  total_mapped: number;
  total_unmapped: number;
  total_verified: number;
  total_kfa_master: number;
  mapped_percentage: number;
  verified_percentage: number;
  ready_for_satusehat: number;
}

export interface UnmappedMedicine {
  id: number;
  code: string;
  name: string;
  generic_name: string;
  form: string;
  strength: string;
  unit: string;
  manufacturer: string;
  category: string;
  type: string;
}

// ============================================================================
// Request Types
// ============================================================================

export interface CreateKFAMasterRequest {
  kfa_code: string;
  kfa_name: string;
  kfa_type?: KFAProductType;
  kfa_farmalkes?: string;
  form?: string;
  strength?: string;
  manufacturer?: string;
  category?: string;
  is_active?: boolean;
}

export interface CreateMedicineKFAMappingRequest {
  medicine_id: number;
  kfa_code: string;
  kfa_name: string;
  kfa_type?: KFAProductType;
  kfa_farmalkes?: string;
  kfa_display_name?: string;
  kfa_form?: string;
  kfa_strength?: string;
  kfa_manufacturer?: string;
  is_verified?: boolean;
  is_active?: boolean;
  notes?: string;
}

// ============================================================================
// KFA Master API
// ============================================================================

export const kfaMasterApi = {
  // Get all KFA master data with filters and pagination
  getAll: (params?: {
    search?: string;
    product_type?: KFAProductType;
    category?: string;
    form?: string;
    is_active?: boolean;
    page?: number;
    limit?: number;
  }) => api.get<{ data: KFAMaster[]; total: number; page: number; limit: number }>('/kfa/master', { params }),

  // Get single KFA master by ID
  getById: (id: number) => api.get<{ data: KFAMaster }>(`/kfa/master/${id}`),

  // Create new KFA master
  create: (data: CreateKFAMasterRequest) =>
    api.post<{ message: string; data: KFAMaster }>('/kfa/master', data),

  // Update KFA master
  update: (id: number, data: CreateKFAMasterRequest) =>
    api.put<{ message: string; data: KFAMaster }>(`/kfa/master/${id}`, data),

  // Delete KFA master
  delete: (id: number) => api.delete<{ message: string }>(`/kfa/master/${id}`),

  // Search KFA master for autocomplete
  search: (query: string) =>
    api.get<{ data: KFAMaster[] }>('/kfa/master/search', { params: { q: query } }),
};

// ============================================================================
// Medicine KFA Mapping API
// ============================================================================

export const kfaMappingApi = {
  // Get all mappings with filters and pagination
  getAll: (params?: {
    search?: string;
    medicine_id?: number;
    is_verified?: boolean;
    is_active?: boolean;
    page?: number;
    limit?: number;
  }) => api.get<{ data: MedicineKFAMapping[]; total: number; page: number; limit: number }>('/kfa/mapping', { params }),

  // Get single mapping by ID
  getById: (id: number) => api.get<{ data: MedicineKFAMapping }>(`/kfa/mapping/${id}`),

  // Get mapping by medicine ID
  getByMedicineId: (medicineId: number) =>
    api.get<{ data: MedicineKFAMapping }>(`/kfa/mapping/medicine/${medicineId}`),

  // Create new mapping with validation
  create: async (data: CreateMedicineKFAMappingRequest) => {
    // Frontend validation
    if (!data.medicine_id) {
      throw new Error('Medicine ID is required');
    }
    if (!data.kfa_code || data.kfa_code.trim() === '') {
      throw new Error('KFA Code cannot be blank');
    }
    if (!data.kfa_name || data.kfa_name.trim() === '') {
      throw new Error('KFA Name cannot be blank');
    }

    return api.post<{ message: string; data: MedicineKFAMapping }>('/kfa/mapping', data);
  },

  // Update mapping
  update: (id: number, data: CreateMedicineKFAMappingRequest) =>
    api.put<{ message: string; data: MedicineKFAMapping }>(`/kfa/mapping/${id}`, data),

  // Delete mapping
  delete: (id: number) => api.delete<{ message: string }>(`/kfa/mapping/${id}`),

  // Verify mapping
  verify: (id: number) =>
    api.post<{ message: string; data: MedicineKFAMapping }>(`/kfa/mapping/${id}/verify`),

  // Send standalone Medication resource to SatuSehat
  sendToSatuSehat: (mappingId: number) =>
    api.post<{
      message: string;
      mapping_id: number;
      medicine_name: string;
      kfa_code: string;
      satusehat_medication_id: string;
      fhir_response: Record<string, unknown>;
    }>(`/kfa/medication/send/${mappingId}`),
};

// ============================================================================
// KFA Statistics API
// ============================================================================

export const kfaStatsApi = {
  // Get mapping statistics
  getStats: () => api.get<KFAMappingStats>('/kfa/stats'),

  // Get unmapped medicines
  getUnmapped: (params?: {
    search?: string;
    category?: string;
    form?: string;
    page?: number;
    limit?: number;
  }) => api.get<{ data: UnmappedMedicine[]; total: number; page: number; limit: number }>('/kfa/unmapped', { params }),
};

// ============================================================================
// KFA Lookup from SatuSehat API
// ============================================================================

export interface KFALookupResult {
  kfa_code: string;
  name: string;
  display_name: string;
  manufacturer: string;
  dosage_form: string;
  farmalkes?: string;
  nie?: string;
}

export const kfaLookupApi = {
  // Lookup KFA by code from SatuSehat API
  lookupByCode: async (code: string) => {
    if (!code || code.trim() === '') {
      throw new Error('KFA code is required');
    }
    try {
      const response = await api.get<{ found: boolean; data: KFALookupResult }>(`/kfa/lookup/code/${code}`);
      return response;
    } catch (error: any) {
      // Provide more helpful error messages
      if (error.response?.status === 404) {
        throw new Error(`KFA code "${code}" not found in SatuSehat`);
      }
      if (error.response?.status === 400) {
        throw new Error(error.response?.data?.error || 'Invalid KFA code format');
      }
      throw error;
    }
  },

  // Search KFA from SatuSehat API
  search: async (query: string) => {
    if (!query || query.trim() === '') {
      return { data: { total: 0, data: [] } };
    }
    if (query.trim().length < 3) {
      throw new Error('Search query must be at least 3 characters');
    }
    try {
      const response = await api.get<{ total: number; data: KFALookupResult[] }>('/kfa/lookup/search', { params: { q: query } });
      return response;
    } catch (error: any) {
      if (error.response?.status === 400) {
        throw new Error(error.response?.data?.error || 'Invalid search query');
      }
      throw error;
    }
  },
};

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Convert KFA lookup result to mapping request format
 * Useful for autofilling form data from SatuSehat API lookup
 */
export function kfaResultToMappingData(
  result: KFALookupResult,
  medicineId: number
): Omit<CreateMedicineKFAMappingRequest, 'medicine_id'> & { medicine_id: number } {
  return {
    medicine_id: medicineId,
    kfa_code: result.kfa_code || '',
    kfa_name: result.name || '',
    kfa_display_name: result.display_name || result.name || '',
    kfa_form: result.dosage_form || '',
    kfa_manufacturer: result.manufacturer || '',
    kfa_farmalkes: result.farmalkes || '',
    kfa_type: 'obat' as KFAProductType,
    is_active: true,
  };
}

/**
 * Validate KFA mapping data before submission
 */
export function validateKFAMapping(data: CreateMedicineKFAMappingRequest): { valid: boolean; errors: string[] } {
  const errors: string[] = [];

  if (!data.medicine_id) {
    errors.push('Medicine ID is required');
  }
  if (!data.kfa_code || data.kfa_code.trim() === '') {
    errors.push('KFA Code cannot be blank');
  }
  if (!data.kfa_name || data.kfa_name.trim() === '') {
    errors.push('KFA Name cannot be blank');
  }

  // Recommended fields
  if (!data.kfa_display_name || data.kfa_display_name.trim() === '') {
    errors.push('Display name is recommended for FHIR resources');
  }
  if (!data.kfa_form || data.kfa_form.trim() === '') {
    errors.push('Dosage form is recommended');
  }
  if (!data.kfa_manufacturer || data.kfa_manufacturer.trim() === '') {
    errors.push('Manufacturer is recommended');
  }

  return {
    valid: errors.length === 0 || errors.every(e => e.includes('recommended')),
    errors,
  };
}
